### ISSUE TITLE
Proposal: do not bind mount /etc/hosts, /etc/hostname and /etc/resolv.conf --net=host or add support for --add-hosts when --net=host

### ISSUE THREAD

**Issue by SeanMooney** [moby/moby#22853]
---
## BUG REPORT INFORMATION

Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Output of `docker version`:**

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:40:36 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:40:36 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 17
 Running: 3
 Paused: 0
 Stopped: 14
Images: 505
Server Version: 1.11.0
Storage Driver: btrfs
 Build Version: Btrfs v3.19.1
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.4.4-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 31.29 GiB
Name: ceph-master
ID: EMX6:4HDA:SBCH:VKGA:5E3M:S3JX:F24B:NTQM:ASFX:DDGG:A57H:GXRR
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical 

**Steps to reproduce the issue:**
1.  docker run --rm -it --net=host busybox mount | grep /etc/hosts
2.  docker run --rm -it --net=host --add-host test:1.2.3.4 busybox grep test /etc/hosts

**Describe the results you received:**
1.  output: "/dev/sdb2 on /etc/hosts type btrfs (rw,seclabel,relatime,ssd,space_cache,subvolid=257,subvol=/root/var/lib/docker/containers/a46f7eb4e109b6736ba18c4a04bff662c85aff594b837b564793b158fd981148/hosts)"
/etc/hosts is a bindmount to a docker managed host file.
1.  output:"docker: Error response from daemon: Conflicting options: custom host-to-IP mapping and the network mode."

**Describe the results you expected:**
1. /etc/host is not a bind mount
or
2. --add-hosts work if docker bind mounts /etc/hosts when --net=hosts

**Additional information you deem important (e.g. issue happens only occasionally):**
binding mounting /etc/hosts is required for docker managed networking but is not a requirement when
when host networking is used. bind mounting /etc/hosts prevents many tools such as sed and ansible which do atomic file updates by creating an temp file and then preform an atomic rename as this action is not supported for bind mounted files.

when in --host=net mode the expected behavior is that all docker networking is disabled for this container and host networking is used. in this case ideally  /etc/hosts, /etc/hostname and /etc/resolv.conf  would not be bind mounted into the container so that they can be configured externally using tools such as sed/ansible. this is a problem when using docker to test
deployment scripts that are intended to run on physical machines where the script is expected to update these files.

**Comment by thaJeztah** [comment]
See a related discussion on this PR https://github.com/docker/docker/pull/22408 (https://github.com/docker/docker/pull/22408#issuecomment-220082584)

**Comment by bboreham** [comment]
Looks like this should be fixed by #22408 (I haven't tried it)

**Comment by pavlix** [comment]
In my opinion bind-mounting `/etc/resolv.conf` is wrong in all cases. You might always want to change the contents of the file which is generally done by *replacing*, not *rewriting*, as rewriting doesn't ensure consistency at all times. In many systems `/etc/resolv.conf` is a symlink and it works well. The best way would be to either bind-mount or copy the docker resolv.conf elsewhere in the system, preferably read-only, and just symlinked.

That way you would (1) allow detection that `/etc/resolv.conf` is still linked to the docker-originated resolv-conf file, (2) allow replacing of `/etc/resolv.conf` properly with another symlink or a new regular file in the standard atomic way, and (3) keep the original contents available.

**Comment by henrikno** [comment]
`docker run --rm -it --net=host --add-host test:1.2.3.4 busybox grep test /etc/hosts` works on 
```
Version:           18.09.0
API version:       1.39

docker run --rm -it --net=host --add-host test:1.2.3.4 busybox grep test /etc/hosts
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
57c14dd66db0: Pull complete
Digest: sha256:7964ad52e396a6e045c39b5a44438424ac52e12e4d5a25d94895f2058cb863a0
Status: Downloaded newer image for busybox:latest
1.2.3.4 test
```

**Comment by thaJeztah** [comment]
Yes, this is fixed since Docker 1.12 through https://github.com/moby/moby/pull/22408 (related: https://github.com/moby/moby/pull/20177 which allowed setting `--hostname` for the container itself)

**Comment by kmahyyg** [comment]
For those man who searched and jumped here, after I check Docker related docs, I have to say, Unfortunately, Docker doesn't support disable this behaviour, you might need to use Podman.

Check: https://github.com/containers/podman/pull/2747
